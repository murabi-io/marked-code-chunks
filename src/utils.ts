const escapeTest = /[&<>"']/;
const escapeReplace = /[&<>"']/g;
const escapeTestNoEncode = /[<>"']|&(?!#?\w+;)/;
const escapeReplaceNoEncode = /[<>"']|&(?!#?\w+;)/g;
const escapeReplacements: Record<string, string> = {
  "&": "&amp;",
  "<": "&lt;",
  ">": "&gt;",
  '"': "&quot;",
  "'": "&#39;",
};

const getEscapeReplacement = (ch: string) => escapeReplacements[ch];

export function escape(html: string, encode: boolean) {
  if (encode) {
    if (escapeTest.test(html)) {
      return html.replace(escapeReplace, getEscapeReplacement);
    }
  } else {
    if (escapeTestNoEncode.test(html)) {
      return html.replace(escapeReplaceNoEncode, getEscapeReplacement);
    }
  }

  return html;
}

export function indentCodeCompensation(raw: string, text: string): string {
  const matchIndentToCode = raw.match(/^(\s+)(?:```)/);

  if (matchIndentToCode === null) {
    return text;
  }

  const indentToCode = matchIndentToCode[1];

  return text
    .split("\n")
    .map((node) => {
      const matchIndentInNode = node.match(/^\s+/);
      if (matchIndentInNode === null) {
        return node;
      }

      const [indentInNode] = matchIndentInNode;

      if (indentInNode.length >= indentToCode.length) {
        return node.slice(indentToCode.length);
      }

      return node;
    })
    .join("\n");
}
