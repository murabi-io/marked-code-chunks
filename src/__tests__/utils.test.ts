import { escape, indentCodeCompensation } from "../utils";

describe("utils", () => {
  describe("escape function", () => {
    describe("given a code", () => {
      it("should escape the code, without encoding", () => {
        const html = `
        <div class="token-line css-k008qs" style="color:rgb(214, 222, 235)">
          <span class="css-15">1</span>
          <span class="token function" style="color:rgb(130, 170, 255)">npm</span>
          <span class="token plain"> </span>
          <span class="token function" style="color:rgb(130, 170, 255)">install</span>
          <span class="token plain"> npm-dts --save-dev</span>
         </div>
      `;
        expect(escape(html, false)).toMatchSnapshot();
      });
      it("should escape the code, with encoding", () => {
        const code = `
        const { build } = require('esbuild')
  
        build({
          entryPoints: ['src/index.ts'],
          outdir: 'dist',
          bundle: true,
        })
      `;
        expect(escape(code, true)).toMatchSnapshot();
      });
    });
  });
  describe("indentCodeCompensation function", () => {
    describe("given a code", () => {
      it("should properly indent the given code", () => {
        const text = `
        const { build } = require('esbuild')
        build({
          entryPoints: ['src/index.ts'],
          outdir: 'dist',
          bundle: true,
        })`;
        const raw = `
        \`\`\`js {cmd=node output=txt modify_source}
              ${text}
        \`\`\`
        `;
        expect(indentCodeCompensation(raw, text)).toMatchSnapshot();
      });
    });
  });
});
