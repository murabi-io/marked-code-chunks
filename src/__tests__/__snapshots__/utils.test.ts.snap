// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils escape function given a code should escape the code, with encoding 1`] = `
"
        const { build } = require(&#39;esbuild&#39;)
  
        build({
          entryPoints: [&#39;src/index.ts&#39;],
          outdir: &#39;dist&#39;,
          bundle: true,
        })
      "
`;

exports[`utils escape function given a code should escape the code, without encoding 1`] = `
"
        &lt;div class=&quot;token-line css-k008qs&quot; style=&quot;color:rgb(214, 222, 235)&quot;&gt;
          &lt;span class=&quot;css-15avlhb&quot;&gt;1&lt;/span&gt;
          &lt;span class=&quot;token function&quot; style=&quot;color:rgb(130, 170, 255)&quot;&gt;npm&lt;/span&gt;
          &lt;span class=&quot;token plain&quot;&gt; &lt;/span&gt;
          &lt;span class=&quot;token function&quot; style=&quot;color:rgb(130, 170, 255)&quot;&gt;install&lt;/span&gt;
          &lt;span class=&quot;token plain&quot;&gt; npm-dts --save-dev&lt;/span&gt;
         &lt;/div&gt;
      "
`;

exports[`utils indentCodeCompensation function given a code should properly indent the given code 1`] = `
"
        const { build } = require('esbuild')
        build({
 entryPoints: ['src/index.ts'],
 outdir: 'dist',
 bundle: true,
        })"
`;
